Make Adaptive Speed more fun

Nabile-Rahmani opened the pull request:
* Fixes pathological cases that would occur way too often for me to ignore: accurately full combo a map, yet the speed would stagnate, or much worse, slow down to a crawl, due to overcorrection/lack of reward for high accuracy.

The thresholded accuracy reward lets the mod perfectly adapt to the player's real abilities: no longer does the target rate go below 1.0 on easy maps due to unstable notes, and the rate change is responsive enough to match one's speed limit to be able to complete songs.

Here's a diff play, with the target rate debug log on the left terminal.

New behaviour (video):

[![osu as mod new replay](https://i.ytimg.com/vi/xTTAwGJdDms/maxresdefault.jpg)](https://youtu.be/xTTAwGJdDms)

Old behaviour (video; playing back the same new replay from above):

[![osu as mod old replay](https://i.ytimg.com/vi/RVW7ua_TB1w/maxresdefault.jpg)](https://youtu.be/RVW7ua_TB1w)

0.63x on a full combo. This isn't a prank. Here's the entire diff against the master branch:

```diff
diff --git a/osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs b/osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs
index e7127abcf0..c4f95050c2 100644
--- a/osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs
+++ b/osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs
@@ -16,6 +16,8 @@
 using osu.Game.Rulesets.Scoring;
 using osu.Game.Rulesets.UI;
 
+using osu.Framework.Logging;
+
 namespace osu.Game.Rulesets.Mods
 {
     public class ModAdaptiveSpeed : Mod, IApplicableToRate, IApplicableToDrawableHitObject, IApplicableToBeatmap, IUpdatableByPlayfield
@@ -257,6 +259,8 @@ private void updateTargetRate()
 
             // Scale the rate adjustment based on consistency
             targetRate = Interpolation.Lerp(targetRate, recentRates.Average(), Math.Abs(consistency) / (recent_rate_count - 1d));
+
+            Logger.Log($"[AS mod] Target rate: {targetRate:0.00} (recent rates: {string.Join(", ", recentRates.Select(r => r.ToString("0.00")))})");
         }
     }
 }
```

TL;DR:

```csharp
var rateChange = (result.HitObject.GetEndTime() - prevEndTime) / (result.TimeAbsolute - prevEndTime);

if (Math.Abs(rateChange - 1.0) < rate_change_threshold) // Reward well-timed results with a speed up.
    return rate_change_on_hit;
```

A couple new constants are introduced, and I wondered for a bit whether we should expose them to users to customise further, but they feel just about right !

```csharp
// Apply a fixed rate change when accurately hitting notes, to counteract overcorrection stagnating or even slowing down an accurate but unstable human player.
private const double rate_change_on_hit = max_allowable_rate_change * 0.95;

// The threshold below which we'll reward the player's high accuracy with a speed up.
private const double rate_change_threshold = 0.05;
```

Shall we expose these options as well ?

* Expose min/max speed options, and increase the max speed's range to expand playable maps for high level players.

* Set the score multiplier to the min speed, since this is now dynamic.

<video src="https://github.com/ppy/osu/assets/15223353/2924b160-9f2f-4c1c-9bcf-dcc3d60f6d94">
</video>

Nabile-Rahmani committed:
Adaptive Speed: Expose min/max speed settings

- Increase max speed range from 2 to 4 to expand playable maps for high level players.

- Score multiplier is now set to the min speed bindable.

Nabile-Rahmani committed:
Adaptive Speed: Thresholded accuracy reward

Fixes pathological cases that would occur way too often for me to ignore: accurately full combo a map, yet the speed would stagnate, or much worse, slow down to a crawl, due to overcorrection/lack of reward for high accuracy.

The thresholded accuracy reward lets the mod perfectly adapt to the player’s real abilities: no longer does the target rate go below 1.0 on easy maps due to unstable notes, and the rate change is responsive enough to match one’s speed limit to be able to complete songs.

Nabile-Rahmani committed:
camelCase

Nabile-Rahmani committed:
Use explicit type

peppy commented:
Based on the description alone, I agree that this makes sense (and the previous behaviour was wtf-level for a player using the mod for the first time).

Nabile-Rahmani committed:
AS mod: Miss rate is based on min allowable rate

`0.9 / 0.95 == 0.947`, close enough to the original 0.95.

It doesn't make any difference with existing replays.

Nabile-Rahmani committed:
AS mod: Extract common rate change multiplier

Nabile-Rahmani committed:
AS mod: Rate multiplier setting

Nabile-Rahmani commented:
> Based on the description alone, I agree that this makes sense (and the previous behaviour was wtf-level for a player using the mod for the first time).

Makes me wonder how many people have been put off by this mod because of this :p

The concept is good, so I wanted to do it justice. You'd think the constant fluctuation would mess too much with your timing, but it surprisingly flows naturally.

Anyhow, added a rate multiplier setting for hits/misses to make things more or less spicier.

It's pretty good if I do say so myself.

![image](https://github.com/ppy/osu/assets/15223353/096763b2-1f4c-49d1-930c-a5a85730793d)

Setting it to zero ignores rate change for hits and misses, so you're left with inaccurate changes, and it still works as you'd expect.

Nabile-Rahmani committed:
AS mod: Score multiplier is the min allowable rate

This is because it is the lowest actual speed the mod will play at.

Defaulting min rate to 1 as the allowable rate is already low enough.

Nabile-Rahmani committed:
AS mod: Flex Abs on clock to fix TrackBass tempo crash on rewind

Use the absolute clock time so that we always move towards targetRate, not away from it (as is the case when rewinding).

This prevents TrackBass from throwing an exception as the tempo approaches the lower bound (when AdjustPitch is set to false).

Nabile-Rahmani commented:
Caught a nasty replay rewind crash: because the elapsed time is negative, the damped interpolation would move away from the target rate.

If adjust pitch is set to false, the Bass track would throw an exception as the tempo approached zero.

```csharp
public void Update(Playfield playfield)
{
    // Use the absolute clock time so that we always move towards targetRate, not away from it (as is the case when rewinding).
    // This prevents TrackBass from throwing an exception as the tempo approaches the lower bound (when AdjustPitch is set to false).
    double elapsedTime = Math.Abs(playfield.Clock.ElapsedFrameTime);

    SpeedChange.Value = Interpolation.DampContinuously(SpeedChange.Value, targetRate, 50, elapsedTime);
}
```

I'm not sure why it didn't crash before this PR (or maybe it did under certain circumstances), because I don't think this argument is ever supposed to be negative.

Reverted objects reintroduce their rate and `targetRate` gets updated accordingly, so in theory it should always tend towards `targetRate`, which the `Abs` accomplishes.

Perhaps higher rate changes pissed off the damping function ? It looks like when `targetRate - SpeedChange > 0.01` it keeps growing larger.

Whatever, it works now, rate values seem consistent moving around the track.

Before the fix:

![image](https://github.com/ppy/osu/assets/15223353/13c94601-0eee-424d-853a-54689bdd60c2)

Nabile-Rahmani committed:
AS mod: Use a power curve on the rate multiplier

The previous linear range would get clamped when the rate multiplier was around 0.90.

Using a power curve makes the setting's lower bound meaningful.

The default value should be 0.5085 to perfectly match previous replays on this branch, but 0.51 is imperceptibly close (rates differ by ~0.01), so we keep the bindable's precision as-is.

```
>>> min_allowable_rate_change / 0.95
0.9473684210526316
>>> pow(min_allowable_rate_change, 0.5085)
0.9478340714032744
>>> pow(min_allowable_rate_change, 0.51)
0.947684286809841

>>> max_allowable_rate_change * 0.95
1.0545
>>> pow(max_allowable_rate_change, 0.5085)
1.0545003657593834
>>> pow(max_allowable_rate_change, 0.51)
1.0546654501917019
```

Nabile-Rahmani committed:
AS mod: Use DT & HT's score multiplier logic

Massively nerfs the mod. RIP.
